sdk/include/ndk/haltypes.h

```diff
@@ -39,61 +162,1474 @@ typedef enum _FIRMWARE_REENTRY
 );
 
 typedef
 NTSTATUS
-(NTAPI *PINSTALL_BUS_HANDLER)(
-    _In_ PBUS_HANDLER Bus
+PROCESSOR_HALT_ROUTINE(
+    _Inout_opt_ PVOID Context
 );
 
+typedef PROCESSOR_HALT_ROUTINE *PPROCESSOR_HALT_ROUTINE;
+
 typedef
 NTSTATUS
-(NTAPI *pHalRegisterBusHandler)(
-    _In_ INTERFACE_TYPE InterfaceType,
-    _In_ BUS_DATA_TYPE ConfigSpace,
-    _In_ ULONG BusNumber,
-    _In_ INTERFACE_TYPE ParentInterfaceType,
-    _In_ ULONG ParentBusNumber,
-    _In_ ULONG ContextSize,
-    _In_ PINSTALL_BUS_HANDLER InstallCallback,
-    _Out_ PBUS_HANDLER *BusHandler
+(NTAPI *pHalProcessorHalt)(
+    _In_ ULONG Flags,
+    _Inout_opt_ PVOID Context,
+    _In_ PPROCESSOR_HALT_ROUTINE Halt
 );
 
 typedef
 VOID
-(NTAPI *pHalSetWakeEnable)(
-    _In_ BOOLEAN Enable
+(NTAPI *pHalPciMarkHiberPhase)(
+    VOID
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalQueryProcessorRestartEntryPoint)(
+    _Out_ PPHYSICAL_ADDRESS EntryPoint
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalRequestInterrupt)(
+    _In_ ULONG Gsiv
 );
 
 typedef
 VOID
-(NTAPI *pHalSetWakeAlarm)(
-    _In_ ULONGLONG AlartTime,
-    _In_ PTIME_FIELDS TimeFields
+(NTAPI *pHalPowerEarlyRestore)(
+    _In_ ULONG Phase
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalUpdateCapsule)(
+    _In_ PVOID CapsuleHeaderArray,
+    _In_ ULONG CapsuleCount,
+    _In_opt_ PHYSICAL_ADDRESS ScatterGatherList
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalQueryCapsuleCapabilities)(
+    _In_ PVOID CapsuleHeaderArray,
+    _In_ ULONG CapsuleCount,
+    _Out_ PULONGLONG MaximumCapsuleSize,
+    _Out_ PULONG ResetType
+);
+
+typedef
+BOOLEAN
+(NTAPI *pHalPciMultiStageResumeCapable)(
+    VOID
 );
 
 typedef
 VOID
-(NTAPI *pHalLocateHiberRanges)(
-    _In_ PVOID MemoryMap
+(NTAPI *pHalDmaFreeCrashDumpRegisters)(
+    _In_ ULONG Phase
 );
 
+typedef
+BOOLEAN
+(NTAPI *pHalAcpiAoacCapable)(
+    VOID
+);
+
+#if (NTDDI_VERSION >= NTDDI_WIN10)
 typedef
 NTSTATUS
-(NTAPI *pHalAllocateMapRegisters)(
-    _In_ PADAPTER_OBJECT AdapterObject,
-    _In_ ULONG Unknown,
-    _In_ ULONG Unknown2,
-    PMAP_REGISTER_ENTRY Registers
+(NTAPI *pHalInterruptSetDestination)(
+    _In_ ULONG Gsiv,
+    _In_ PINTERRUPT_VECTOR_DATA VectorData,
+    _In_ PGROUP_AFFINITY TargetProcessors
+);
+#else
+typedef
+NTSTATUS
+(NTAPI *pHalInterruptSetDestination)(
+    _In_ PINTERRUPT_CONNECTION_DATA ConnectionData,
+    _In_ PGROUP_AFFINITY TargetProcessors
+);
+#endif
+
+typedef union _HAL_UNMASKED_INTERRUPT_FLAGS
+{
+    struct {
+        USHORT SecondaryInterrupt: 1;
+        USHORT Reserved: 15;
+    };
+    USHORT AsUSHORT;
+} HAL_UNMASKED_INTERRUPT_FLAGS, *PHAL_UNMASKED_INTERRUPT_FLAGS;
+
+typedef struct _HAL_UNMASKED_INTERRUPT_INFORMATION
+{
+    USHORT Version;
+    USHORT Size;
+    HAL_UNMASKED_INTERRUPT_FLAGS Flags;
+    KINTERRUPT_MODE Mode;
+    KINTERRUPT_POLARITY Polarity;
+    ULONG Gsiv;
+    USHORT PinNumber;
+    PVOID DeviceHandle;
+} HAL_UNMASKED_INTERRUPT_INFORMATION, *PHAL_UNMASKED_INTERRUPT_INFORMATION;
+
+typedef
+BOOLEAN
+(NTAPI HAL_ENUMERATE_INTERRUPT_SOURCE_CALLBACK)(
```

Comment:
This is not defined directly as: `... (NTAPI *PHAL_ENUMERATE_INTERRUPT_SOURCE_CALLBACK)(...);` ?

Reply:
Yeah, this is how it is defined in WDK.

Reply:
> Yeah, this is how it is defined in WDK.

You mean, the `HAL_ENUMERATE_INTERRUPT_SOURCE_CALLBACK` (without "pointer") and a separate typedef for `PHAL_ENUMERATE_INTERRUPT_SOURCE_CALLBACK` afterwards? If so this is strange (by that I mean, not the usual way).

Reply:
> You mean, the HAL_ENUMERATE_INTERRUPT_SOURCE_CALLBACK (without "pointer") and a separate typedef for PHAL_ENUMERATE_INTERRUPT_SOURCE_CALLBACK afterwards?

Yes.

Reply:
Let's then keep the WDK way of splitting the "non-pointer" and "pointer" variations.